### PR TITLE
Implementing smbclient collector

### PR DESCRIPTION
**Pull request by rob-scheepens**
This request is to implement a collector for smb-client metrics (performance object "SMB Client Shares").

### CODE REVIEW

File: docs/collector.smbclient.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+Comma-separated list of collectors to use, for example: `--collectors.smbclient.enabled=ServerShares`. Matching is case-sensitive. Depending on the smb protocol version not all performance counters may be available. Use `--collectors.smbclient.list` to obtain a list of supported collectors.
+
+## Metrics
+Name          | Description
+--------------|---------------
+`windows_smbclient_client_shares_avg_sec_per_read` | The average latency between the time a read request is sent and when its response is received.
+`windows_smbclient_client_shares_avg_sec_per_write` | The average latency between the time a write request is sent and when its response is received.
+
+### Example metric
+_This collector does not yet have explained examples, we would appreciate your help adding them!_
```
Comment (jkroepke) [inline docs/collector.smbclient.md:28]: Could you please add `Example metric`?

It would be helpful as part of the review here, too

Comment (rob-scheepens) [reply docs/collector.smbclient.md]: Added in 136308981dccf8777eebfcfa3fe5a114c1e20e08.

File: .golangci.yaml
Comment (jkroepke) [inline .golangci.yaml:1]: Please revert the changes here. Changes on the linter should be done and discussed in a Separate PR

File: docs/collector.smbclient.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+## Metrics
+Name          | Description
+--------------|---------------
+`windows_smbclient_client_shares_avg_sec_per_read` | The average latency between the time a read request is sent and when its response is received.
+`windows_smbclient_client_shares_avg_sec_per_write` | The average latency between the time a write request is sent and when its response is received.
+
+### Example metric
+windows_smbclient_read_bytes_total - The bytes read on this share.
+windows_smbclient_write_bytes_total - The bytes written on this share.
```
Comment (jkroepke) [inline docs/collector.smbclient.md:29]: Please provider examples from the `/metrics` endpoint, including all metrics and labels

File: docs/collector.smbclient.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+## Flags
+
+### `--collectors.smbclient.list`
+Lists the Perflib Objects that are queried for data along with the perlfib object id
+
+### `--collectors.smbclient.enabled`
+Comma-separated list of collectors to use, for example: `--collectors.smbclient.enabled=ServerShares`. Matching is case-sensitive. Depending on the smb protocol version not all performance counters may be available. Use `--collectors.smbclient.list` to obtain a list of supported collectors.
+
+## Metrics
```
Comment (jkroepke) [inline docs/collector.smbclient.md:21]: I have the feeling that this feature list is not complete.

File: pkg/collector/smbclient/smbclient.go [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+	// 	"bytes_per_read",
+	// 	"The average number of bytes per read request.",
+	//   []string{"server", "share"},
+	// 	)
+	// c.AvgBytesPerWrite = desc("bytes_per_write",
+	// 	"The average number of bytes per write request.",
+	//   []string{"server", "share"},)
+	// c.AvgDataBytesPerRequest = desc("data_bytes_per_request",
+	// 	"The average number of bytes per read or write request.",
+	//   []string{"server", "share"},)
```
Comment (jkroepke) [inline pkg/collector/smbclient/smbclient.go:143]: remove comment out code or include it

**Review by jkroepke — CHANGES_REQUESTED** [review]

File: pkg/collector/smbclient/smbclient.go (hunk 4 above)
Comment (ballegre) [reply pkg/collector/smbclient/smbclient.go:143]: Hi I'm Bob. I work with Rob on windows exporter. 

That's my code, I will be happy to remove the commented out code. I would like to include any counter considered useful, but my experience with SMB is limited to the workloads we are running now (simple read and write). I was curious to know if there is anyway to find out more about the perflib or counter definition. I read the comments in perflib package. The closest I came was using `Get-Counter` commands but didn't see any help for a specific counter, just the counter set. Is there a way to get more information about the definition of a counter? Otherwise, I guess it's just a best-effort process. Thanks.

Comment (jkroepke) [reply pkg/collector/smbclient/smbclient.go:143]: Site Note: In my eye, comment code is dead code and mostly will not survive any refactoring.


About the Perf Counter questions:

The GUI Tool "Performance Monitor" is a builtin tool (perfmon.msc) to obverse and see the existing Performance Counters in a System. 

See also: https://www.progress.com/blogs/windows-performance-counters-getting-started

File: docs/collector.smbclient.md (hunk 2 above)
Comment (ballegre) [reply docs/collector.smbclient.md:29]: Is this what you want? These are all the `windows_smbclient` metrics produced on a scrape.
```
# HELP windows_smbclient_current_data_queues The point in time number of requests outstanding on this share.
# TYPE windows_smbclient_current_data_queues counter
windows_smbclient_current_data_queues{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_current_data_queues{server="localhost",share="c$"} 0
# HELP windows_smbclient_data_bytes_total The bytes read or written on this share.
# TYPE windows_smbclient_data_bytes_total counter
windows_smbclient_data_bytes_total{server="10.57.120.60",share="genshare0"} 2.19776e+07
windows_smbclient_data_bytes_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_metadata_requests_total The metadata requests on this share.
# TYPE windows_smbclient_metadata_requests_total counter
windows_smbclient_metadata_requests_total{server="10.57.120.60",share="genshare0"} 30
windows_smbclient_metadata_requests_total{server="localhost",share="c$"} 167
# HELP windows_smbclient_read_bytes_total The bytes read on this share.
# TYPE windows_smbclient_read_bytes_total counter
windows_smbclient_read_bytes_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_read_bytes_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_read_bytes_via_smbdirect_total The bytes read from this share via RDMA direct placement.
# TYPE windows_smbclient_read_bytes_via_smbdirect_total counter
windows_smbclient_read_bytes_via_smbdirect_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_read_bytes_via_smbdirect_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_read_requests_total The read requests on this share.
# TYPE windows_smbclient_read_requests_total counter
windows_smbclient_read_requests_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_read_requests_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_read_requests_via_smbdirect_total The read requests on this share via RDMA direct placement.
# TYPE windows_smbclient_read_requests_via_smbdirect_total counter
windows_smbclient_read_requests_via_smbdirect_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_read_requests_via_smbdirect_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_read_seconds_total Seconds waiting for read requests on this share.
# TYPE windows_smbclient_read_seconds_total counter
windows_smbclient_read_seconds_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_read_seconds_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_reads_queued The number of read requests queued on this share.
# TYPE windows_smbclient_reads_queued counter
windows_smbclient_reads_queued{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_reads_queued{server="localhost",share="c$"} 0
# HELP windows_smbclient_request_seconds_total Seconds waiting for requests on this share.
# TYPE windows_smbclient_request_seconds_total counter
windows_smbclient_request_seconds_total{server="10.57.120.60",share="genshare0"} 32.2819863
windows_smbclient_request_seconds_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_requests_queued_total The number of requests queued on this share.
# TYPE windows_smbclient_requests_queued_total counter
windows_smbclient_requests_queued_total{server="10.57.120.60",share="genshare0"} 3.22819863e+08
windows_smbclient_requests_queued_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_requests_total The requests on this share.
# TYPE windows_smbclient_requests_total counter
windows_smbclient_requests_total{server="10.57.120.60",share="genshare0"} 21
windows_smbclient_requests_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_stalls_total The number of requests delayed based on insufficient credits on this share.
# TYPE windows_smbclient_stalls_total counter
windows_smbclient_stalls_total{server="10.57.120.60",share="genshare0"} 3
windows_smbclient_stalls_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_turbo_io_reads_total The read requests that go through Turbo I/O
# TYPE windows_smbclient_turbo_io_reads_total counter
windows_smbclient_turbo_io_reads_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_turbo_io_reads_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_turbo_io_writes_total The write requests that go through Turbo I/O
# TYPE windows_smbclient_turbo_io_writes_total counter
windows_smbclient_turbo_io_writes_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_turbo_io_writes_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_write_bytes_total The bytes written on this share.
# TYPE windows_smbclient_write_bytes_total counter
windows_smbclient_write_bytes_total{server="10.57.120.60",share="genshare0"} 2.19776e+07
windows_smbclient_write_bytes_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_write_bytes_via_smbdirect_total The written bytes to this share via RDMA direct placement.
# TYPE windows_smbclient_write_bytes_via_smbdirect_total counter
windows_smbclient_write_bytes_via_smbdirect_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_write_bytes_via_smbdirect_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_write_requests_total The write requests on this share.
# TYPE windows_smbclient_write_requests_total counter
windows_smbclient_write_requests_total{server="10.57.120.60",share="genshare0"} 21
windows_smbclient_write_requests_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_write_requests_via_smbdirect_total The write requests to this share via RDMA direct placement.
# TYPE windows_smbclient_write_requests_via_smbdirect_total counter
windows_smbclient_write_requests_via_smbdirect_total{server="10.57.120.60",share="genshare0"} 0
windows_smbclient_write_requests_via_smbdirect_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_write_seconds_total Seconds waiting for write requests on this share.
# TYPE windows_smbclient_write_seconds_total counter
windows_smbclient_write_seconds_total{server="10.57.120.60",share="genshare0"} 32.2819863
windows_smbclient_write_seconds_total{server="localhost",share="c$"} 0
# HELP windows_smbclient_writes_queued The number of write requests queued on this share.
# TYPE windows_smbclient_writes_queued counter
windows_smbclient_writes_queued{server="10.57.120.60",share="genshare0"} 3.22819863e+08
windows_smbclient_writes_queued{server="localhost",share="c$"} 0
```

Comment (jkroepke) [reply docs/collector.smbclient.md:29]: Yes!

File: pkg/collector/smbclient/smbclient.go [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+	WriteRequestsPerSec                        float64 `perflib:"Write Requests/sec"`
+}
+
+func (c *collector) collectClientShares(ctx *types.ScrapeContext, ch chan<- prometheus.Metric) error {
+	var data []perflibClientShares
+	if err := perflib.UnmarshalObject(ctx.PerfObjects["SMB Client Shares"], &data, c.logger); err != nil {
+		return err
+	}
+	for _, instance := range data {
+		// labelName := c.toLabelName(instance.Name)
```
Comment (ballegre) [inline pkg/collector/smbclient/smbclient.go:288]: What should we do about this commented out line and the unused func? I don't see any prometheus restrictions on label values. I looked here https://prometheus.io/docs/practices/naming/.

Comment (jkroepke) [reply pkg/collector/smbclient/smbclient.go:288]: I'm also not aware of any restrictions here.

File: docs/collector.smbclient.md (hunk 3 above)
Comment (ballegre) [reply docs/collector.smbclient.md:21]: The following table represents what I have in the code at this point. The metrics marked TBD have yet to be verified, or I don't understand the data I'm seeing.  I'm interested in the queue metrics, but as I said having a hard time figuring out what the data is showing me.

Name | Description | Type | Labels
-----|-------------|------|-------
`requests_queued_total` | The number of requests queued on this share. | TBD | `server`, `share` |
`reads_queued` | The number of read requests queued on this share. | TBD | `server`, `share` |
`writes_queued` | The number of write requests queued on this share. | TBD | `server`, `share` |
`request_seconds_total` | Seconds waiting for requests on this share. | counter | `server`, `share` |
`stalls_total` | The number of requests delayed based on insufficient credits on this share. | TBD | `server`, `share` |
`current_data_queues` | The point in time number of requests outstanding on this share. | TBD | `server`, `share` |
`data_bytes_total` | The bytes read or written on this share. | counter | `server`, `share` |
`requests_total` | The requests on this share. | counter | `server`, `share` |
`metadata_requests_total` | The metadata requests on this share. | counter | `server`, `share` |
`read_bytes_via_smbdirect_total` | The bytes read from this share via RDMA direct placement. | TBD | `server`, `share` |
`read_bytes_total` | The bytes read on this share. | counter | `server`, `share` |
`read_requests_via_smbdirect_total` | The read requests on this share via RDMA direct placement. | TBD | `server`, `share` |
`read_requests_total` | The read requests on this share. | counter | `server`, `share` |
`turbo_io_reads_total` | The read requests that go through Turbo I/O | TBD | `server`, `share` |
`turbo_io_writes_total` | The write requests that go through Turbo I/O | TBD | `server`, `share` |
`write_bytes_via_smbdirect_total` | The written bytes to this share via RDMA direct placement. | TBD | `server`, `share` |
`write_bytes_total` | The bytes written on this share. | counter | `server`, `share` |
`write_requests_via_smbdirect_total` | The write requests to this share via RDMA direct placement. | TBD | `server`, `share` |
`write_requests_total` | The write requests on this share. | counter | `server`, `share` |
`read_seconds_total` | Seconds waiting for read requests on this share. | counter | `server`, `share` |
`write_seconds_total` | Seconds waiting for write requests on this share. | counter | `server`, `share` |

Comment (ballegre) [reply docs/collector.smbclient.md:21]: I will add `windows_smbclient` prefix.

> Based on the source code, everything is marked as counter. Please validate this.

By source code, do you mean `smbclient.go` source? 
Not every metric I added ends with `_total` suffix. For example `writes_queued` where perflib counter name is `perflib:"Avg. Write Queue Length"`. Because it says `Avg.`, does that make it a counter? 

There are 2 situations:
1. The metrics are zero and I cannot tell if they are incrementing. Many of those are marked TBD.
2. I can see the metrics increment and they don't make sense to me. Also marked TBD.

In case 1, we can guess or we can leave them out.
In case 2, I can mark them as counter, but the description will only be a guess and could be wrong. Is it possible to ask MSFT? Has anyone from perf team at MSFT contributed here?

And yes, `current_data_queues`,  `perflib:"Current Data Queue Length"`, is indeed a gauge. I just verified. The value aligns with the number of outstanding I/O 's in our workload generators so it appears to be requests queued. BTW, the metric name doesn't seem right. I will change it to `current_data_queue_length` to be consistent with perflib name. 

The remaining "queue" counters I cannot figure out (for example `perflib:"Avg. Data Queue Length"` I named `requests_queued_total`). I can see counter incrementing. The rate per second increase is in the millions. Way too high to make sense. Is the value bytes, or milliseconds? There's no clue in the name so I don't know.

Comment (rob-scheepens) [reply docs/collector.smbclient.md:21]: Hi Bob, on https://barreto.home.blog/2012/11/19/windows-server-2012-file-server-tip-new-per-share-smb-client-performance-counters-provide-great-insight/ there is a table comparing smb counters with their physical disk counterparts. Perhaps that helps explaining what they are.

Avg. Data Queue Length in perfmon is described as "The average number of both read and write requests that were queued for this share.", so not the number of bytes, but number of requests. But, millions per second indeed is odd.

Comment (ballegre) [reply docs/collector.smbclient.md:21]: Also, the exporter converts `Avg. sec/Data Request` from ticks to seconds before it is exposed. The `Avg. Data Queue Length` is exposed raw so the conversion to seconds already has happened.

File: pkg/collector/smbclient/smbclient.go (hunk 5 above)
Comment (ballegre) [reply pkg/collector/smbclient/smbclient.go:288]: Yes. I am currently running with it commented out.

File: docs/collector.smbclient.md (hunk 3 above)
Comment (ballegre) [reply docs/collector.smbclient.md:21]: I checked to see how the physical disk collector handles the queue len counters. It simply doesn't expose them. The only queue counter exposed is:
```
type PhysicalDisk struct {
	Name                   string
	CurrentDiskQueueLength float64 `perflib:"Current Disk Queue Length"`
```
This is the same gauge that we already verified.

Comment (ballegre) [reply docs/collector.smbclient.md:21]: I may have figured it out. I think the value provided by `perflib: "Avg. Data Queue Length"` unit is milliseconds of wait time, represented as ticks (1/10,000). So to convert to seconds, as required by the exporter, we multiply the value by `perflib.TicksToSecondScaleFactor` to convert from ticks to milliseconds, and then multiply by 1/1000 (or millisecond scale factor) to convert from milliseconds to seconds.

We then use wait seconds from above to calculate queue length using Little's Law:
```
  Q = R * W, where,
  Q is the queue len,
  R is the request rate,
  W is wait seconds from above
```
I checked this against performance tests using 5 workloads where we know the number of outstanding IOs in advance (labeled the expected value in table). These were the results:
workload| expected | actual |
-|-|-
prefill seq write | 32 | 31.7 |
rand read | 32 | 28.7 |
seq read | 16 | 14.9 |
seq write | 8 | 7.87 |
rand write | 32 | 31.3 |

Comment (ballegre) [reply docs/collector.smbclient.md:21]: Also, completely independently, on an entirely different test bed, our SMB performance analyst (not me) confirmed that `perfmon` shows the `Avg. Data Queue Length` equal to 31.82 on `rand write` workload when the expected value is 32. So this is evidence that my queue length calculations align with `perfmon`.

Comment (ballegre) [reply docs/collector.smbclient.md:17]: Documenting the Prometheus expression I currently use to get the queue length. Important to note that the metric is the raw value returned from `perflib:"Avg. Data Queue Length"`, the exporter metric name is tentative and could change.
```
irate(windows_smbclient_requests_queued_total[4s]) * (1/10000000)
```
The confusing part is that when consider the counter as ticks (time) and convert to seconds, it produces the queue length (not a time value).

**Comment by ballegre** [comment]
I'm done with the code changes and testing that I can think of at this stage. This is my first contribution to a github repo. I can squash my commits to a single commit. Maybe I can squash Robs commits as well if Rob is OK with that and git will let me. Please let me know what is required or needed next. Thanks.

File: .golangci.yaml [hunk 6]
```diff
@@ -1,14 +1,13 @@
 linters:
   disable-all: true
   enable:
-  - deadcode
   - errcheck
```
Comment (jkroepke) [inline .golangci.yaml:4]:
```suggestion
  - deadcode
  - errcheck
```

File: .golangci.yaml [hunk 7]
```diff
@@ -1,14 +1,13 @@
...
   - revive
   - govet
   - gofmt
   - ineffassign
   - unconvert
-  - varcheck
+  - unused
```
Comment (jkroepke) [inline .golangci.yaml:10]:
```suggestion
  - varcheck
```

**Review by jkroepke — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -49,6 +49,7 @@ Name     | Description | Enabled by default
 [scheduled_task](docs/collector.scheduled_task.md) | Scheduled Tasks metrics |
 [service](docs/collector.service.md) | Service state metrics | &#10003;
 [smb](docs/collector.smb.md) | SMB Server |
+[smbclient](docs/collector.smbclient.md) | SMB Server |
```
Comment (jkroepke) [inline README.md:52]:
```suggestion
[smbclient](docs/collector.smbclient.md) | SMB Client |
```

**Comment by jkroepke** [comment]
Thanks!
